vllm/v1/attention/backends/flex_attention.py

```diff
@@ -0,0 +1,481 @@
+flex_attention_compiled = torch.compile(flex_attention, fullgraph=True)
+
+
+class FlexAttentionBackend(AttentionBackend):
+    accept_output_buffer: bool = True
+
+    @staticmethod
+    def get_supported_head_sizes() -> list[int]:
+        return [32, 64, 96, 128, 160, 192, 224, 256]
+
+    @staticmethod
+    def get_name() -> str:
+        return "FLEX_ATTENTION_VLLM_V1"
+
+    @staticmethod
+    def get_impl_cls() -> type["FlexAttentionImpl"]:
+        return FlexAttentionImpl
+
+    @staticmethod
+    def get_metadata_cls() -> type["AttentionMetadata"]:
+        return FlexAttentionMetadata
+
+    @staticmethod
+    def get_kv_cache_shape(
+        num_blocks: int,
+        block_size: int,
+        num_kv_heads: int,
+        head_size: int,
+    ) -> tuple[int, ...]:
+        if block_size % 16 != 0:
+            raise ValueError("Block size must be a multiple of 16.")
+        return (2, num_blocks, block_size, num_kv_heads, head_size)
+
+    @staticmethod
+    def get_builder_cls() -> type["FlexAttentionMetadataBuilder"]:
+        return FlexAttentionMetadataBuilder
+
+    @staticmethod
+    def use_cascade_attention(*args, **kwargs) -> bool:
+        return False
+
+
+# @torch.compile(fullgraph=True)
+def physical_to_logical_mapping(
+        block_table: torch.Tensor,
+        total_blocks: Optional[int] = None) -> torch.Tensor:
+    """
+    Creates an inverse mapping from physical block locations to logical indices.
+
+    The original block_table maps from logical blocks to physical locations:
+
+    Logical to Physical (Original block_table):
+    ┌───────────────────────────────────────────┐
+    │ Request 0:                                │
+    │                                           │
+    │ Logical Blocks:  0  1  2  3  4  5  6  7   │
+    │                  │  │  │  │  │  │  │  │   │
+    │                  v  v  v  v  v  v  v  v   │
+    │ Physical Blocks: 3  5  1  7  4  2  0  6   │
+    └───────────────────────────────────────────┘
+
+    This function creates the inverse mapping:
+
+    Physical to Logical (Inverse mapping):
+    ┌───────────────────────────────────────────┐
+    │ Request 0:                                │
+    │                                           │
+    │ Physical Blocks: 0  1  2  3  4  5  6  7   │
+    │                  │  │  │  │  │  │  │  │   │
+    │                  v  v  v  v  v  v  v  v   │
+    │ Logical Blocks:  6  2  5  0  4  1  7  3   │
+    └───────────────────────────────────────────┘
+
+    If multiple logical blocks map to the same physical block,
+    this function returns the first (minimum) logical block index.
+
+    If a physical block is not mapped to by any logical block,
+    its value in the result will be -1.
+
+
+    Args:
+        block_table: Tensor of shape [max_reqs, max_num_blocks]
+            mapping logical blocks to physical locations
+
+    Returns:
+        A tensor of shape [max_reqs, max_physical_block]
+    """
```

Comment:
@tlrmchlsmth @chanh FlexAttention is CUDAGraphable so we could also use full CUDAGraphs with it. Is there anything special we would need to do with the unified_attention operator (i.e. is the non-cudagraphable part of the unified_attention operator just dependent on the attention backend?)

Reply:
Agree there is some overlap here obviously: https://github.com/vllm-project/vllm/pull/16072

Reply:
Sorry missed this -- There shouldn't be anything special needed, and I think the only change we need to make is to enable it here: https://github.com/vllm-project/vllm/blob/a944f8ede7361a5233e112a575ff77c4aaa268a5/vllm/v1/worker/gpu_model_runner.py#L143-L151

Reply:
Is this meaningfully different then: https://github.com/vllm-project/vllm/pull/16078/files#diff-3bb4c705122f0ac754ce571780b1d75d5a8d444b2c99581385c1adfaeb8562e2R145
?

Reply:
I think we should add either a property or likely better a method `supports_cuda_graph` to the attention backend that we call in the code @tlrmchlsmth linked above. Then that gets checked in the runner or even better config, if possible.

We should likely also do an audit for other backends - I assume at least the Triton backends could be made to support full CUDA graphs?

Reply:
> Is this meaningfully different then: https://github.com/vllm-project/vllm/pull/16078/files#diff-3bb4c705122f0ac754ce571780b1d75d5a8d444b2c99581385c1adfaeb8562e2R145 ?

I think yes - FA3 supports full CUDA graphs even though it doesn't use a direct call. But correct me if I'm wrong here